patches/ui-webui-resources-tools-build_webui.gni.patch

```diff
@@ -1,28 +1,52 @@
+index f64b09894c1ae..3dee5232a902a 100644
 --- a/ui/webui/resources/tools/build_webui.gni
 +++ b/ui/webui/resources/tools/build_webui.gni
-@@ -190,6 +190,7 @@ template("build_webui") {
+@@ -111,6 +111,7 @@ template("build_webui") {
+       css_wrapper_files += [ f + ".ts" ]
+     }
+   }
++  override_output_files = [] if (defined(invoker.override_files)) { foreach(override, invoker.override_files) { override_output_files += [get_path_info(override, "dir") + "/" + get_path_info(override, "name") + "-chromium." + get_path_info(override, "extension") ] } } override_ts_files = filter_include(override_output_files, ["*.ts"]) if (defined(invoker.css_files)) { override_css_files = filter_include(override_output_files, ["*.css"]) foreach(f, override_css_files) { css_wrapper_files += [f + ".ts" ] } }
+ 
+   # Generated Mojo JS files.
+   if (defined(invoker.mojo_files)) {
+@@ -190,6 +191,7 @@ template("build_webui") {
          enable_removal_comments = enable_source_maps
        }
        in_files = ts_files
-+      if (defined(invoker.exclude_ts_preprocess_files)) { in_files -= invoker.exclude_ts_preprocess_files } if (defined(invoker.preprocess_deps)) { public_deps = invoker.preprocess_deps }
++      if (defined(invoker.override_files)) {         override_files = filter_include(invoker.override_files + invoker.replace_files, ["*.ts"]) override_file_outputs = override_ts_files } if (defined(invoker.exclude_ts_preprocess_files)) { in_files -= invoker.exclude_ts_preprocess_files } if (defined(invoker.preprocess_deps)) { public_deps = invoker.preprocess_deps } 
      }
    }
  
-@@ -240,6 +241,7 @@ template("build_webui") {
+@@ -240,6 +242,7 @@ template("build_webui") {
        if (defined(invoker.css_files)) {
          in_files += invoker.css_files
        }
-+      if (defined(invoker.exclude_html_css_preprocess_files)) { in_files -= invoker.exclude_html_css_preprocess_files } if (defined(invoker.preprocess_deps)) { public_deps = invoker.preprocess_deps }
++      if (defined(invoker.override_files)) { override_files = filter_include(invoker.override_files + invoker.replace_files, ["*.css"]) override_file_outputs = override_css_files } if (defined(invoker.exclude_html_css_preprocess_files)) { in_files -= invoker.exclude_html_css_preprocess_files } if (defined(invoker.preprocess_deps)) { public_deps = invoker.preprocess_deps }
      }
    }
  
-@@ -266,6 +268,7 @@ template("build_webui") {
+@@ -266,6 +269,7 @@ template("build_webui") {
            }
          }
        }
 +      if (defined(invoker.preprocess_deps)) { public_deps = invoker.preprocess_deps }
      }
    }
  
+@@ -276,6 +280,7 @@ template("build_webui") {
+       in_folder = preprocess_dir
+       out_folder = preprocess_dir
+       in_files = invoker.css_files
++      in_files += override_css_files
+       minify = optimize
+     }
+   }
+@@ -350,6 +355,7 @@ template("build_webui") {
+     }
+ 
+     in_files = ts_files
++    in_files += override_ts_files
```

Comment:
there's a lot of patches in this approach. Have you considered using `chromium_src` dir instead of manually passing `--overrides` and `--override-in-folder`? Maybe it's possible to lookup the override files using the input files list you have? Something similar to what I did here: https://github.com/brave/brave-core/pull/25023

You can `touch` original files when `chromium_src` counterpart is modified by extending this list in `util.js`:
```
    // Return true when original file of |file| should be touched.
    const applyFileFilter = (file) => {
      // Only include overridable files.
      const supportedExts = ['.cc', '.h', '.icon', '.json', '.mm', '.mojom', '.pdl', '.py', '.ts'];
      return supportedExts.includes(path.extname(file))
    }
```

This way you don't need to add those files as GN deps to trigger a rebuild on change.

Reply:
I don't quite get the difference between `override` and `replace` we have here. Override is like "keep the original, but also add this", and replace is "replace the original entirely"?

Reply:
You can run `main()` two times with your own arguments the second time to pass custom `out-folder`, `in-folder` and `in-files` vars, then you can merge/rewrite resulted files the way you want.

Reply:
I actually set something similar to that up in `chromium_src` earlier, but when I was talking to @bridiver about it he indicated he'd prefer something a bit more explicit. I'm happy to move back towards that approach, because I think it makes overrides easier (and also, it's a lot more similar to how we override C++ files).

Yup, that's the difference between override and replace. I separated them because we need to know whether to include the original file in the Typescript build (and we can't always include both because we'll get mismatched type definitions).

Calling `main()` twice is genius!

Reply:
I don't recall the original conversation, can you link to it?

Reply:
@fallaciousreasoning looking at the PR @goodov is referencing, I think we're talking about different things here in terms of chromium src. If I recall correctly you were doing overrides to generated files I think? Something like what @goodov has done in his PR seems fine to me

Reply:
maybe I'm misremembering though so a link to the previous revision of this or the discussion would be helpful

Reply:
I am sure I did not have something like this in mind where you would be adding options to the script https://github.com/brave/brave-core/pull/26333/files#diff-5521e907eb8b28af32aec737b47fe8aaae90e847310a25515e439c2350428dd0R9

Reply:
I'll try and dig it up - happy to go the chromium_src route here.

FWIW, this is also overriding generated files, as the `preprocess_if_expr` target is the easiest place for us to copy/modify files :smile:

Reply:
What do you guys think of this? 
https://github.com/brave/brave-core/pull/26676

I've converted it over to a `chromium_src` based approach and it definitely works a bit nicer (there's a lot less patching).

@goodov I haven't migrated the `main` function over yet but happy to do that after you guys have a look at the general approach.

Reply:
> What do you guys think of this? #26676
> 
> I've converted it over to a `chromium_src` based approach and it definitely works a bit nicer (there's a lot less patching).
> 
> @goodov I haven't migrated the `main` function over yet but happy to do that after you guys have a look at the general approach.

This does look much cleaner, I'm in favor of this approach.